adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,257 @@
+        # by way of its middleware system
+        #
+        # Notable implementation differences from dd-trace-rb:
+        # * missing: 'span.resource', which is set to span.name
+        # * missing: config[:distributed_tracing]
+        # * missing: span.set_error() -- spec level change
+        class TracerMiddleware # rubocop:disable Metrics/ClassLength
+          class << self
+            def allowed_rack_request_headers
+              @allowed_rack_request_headers ||= Array(config[:allowed_request_headers]).each_with_object({}) do |header, memo|
+                memo["HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"] = build_attribute_name('http.request.headers.', header)
+              end
+            end
+
+            def allowed_response_headers
+              @allowed_response_headers ||= Array(config[:allowed_response_headers]).each_with_object({}) do |header, memo|
+                memo[header] = build_attribute_name('http.response.headers.', header)
+                memo[header.to_s.upcase] = build_attribute_name('http.response.headers.', header)
+              end
+            end
+
+            def build_attribute_name(prefix, suffix)
+              prefix + suffix.to_s.downcase.gsub(/[-\s]/, '_')
+            end
+
+            def config
+              Rack::Adapter.instance.config
+            end
+
+            private
+
+            def clear_cached_config
+              @allowed_rack_request_headers = nil
+              @allowed_response_headers = nil
+            end
+          end
+
+          EMPTY_HASH = {}.freeze
+
+          def initialize(app)
+            @app = app
+          end
+
+          def call(env) # rubocop:disable Metrics/AbcSize, Metrics/MethodLength
+            original_env = env.dup
+
+            extracted_context = OpenTelemetry.propagation.extract(env)
+
+            # restore extracted context in this process:
+            OpenTelemetry::Context.with_current(extracted_context) do
+              frontend_span = create_frontend_span(env)
+
+              # http://www.rubydoc.info/github/rack/rack/file/SPEC
+              # The source of truth in Rack is the PATH_INFO key that holds the
+              # URL for the current request; but some frameworks may override that
+              # value, especially during exception handling.
+              #
+              # Because of this, we prefer to use REQUEST_URI, if available, which is the
+              # relative path + query string, and doesn't mutate.
+              #
+              # REQUEST_URI is only available depending on what web server is running though.
+              # So when its not available, we want the original, unmutated PATH_INFO, which
+              # is just the relative path without query strings.
+              request_span_name = create_request_span_name(env['REQUEST_URI'] || original_env['PATH_INFO'])
+
+              rack_request = ::Rack::Request.new(env)
```

Comment:
We can avoid allocating the Rack::Request as it's just a wrapper around the `env`. It looks like we have wrapper methods on top of that and some backwards compatibility special handling to top it off. If it's not too much trouble to work on the env itself, we could yak shave an allocation.

Reply:
I should be completely clear that, if we are using some of the convenience methods and it would be hard to live without them, we can leave this as is. Only yak shave if it's easy.

Reply:
After removing the extra attributes (`request_span_attributes`), I was able to inline the leftover `Rack::Request#fullpath` functionality without too much difficulty.